### PR TITLE
Infer from generic function return types

### LINKED ISSUE #15680: Smarter type inference for functional programming (RxJS, Ramda, etc)
**EDIT: updated the types in example to match what works perfectly in Flow type**

**TypeScript Version:**  2.2.1 / nightly (2.2.0-dev.201xxxxx)

TypeScript is unable to infer types through higher-order functions.

**Code**

Given some imaginary class `SetOf`, which is just some set of values we want to compose operations over... We'll try the following:

```ts
// This is a contrived class. We could do the same thing with Observables, etc.
class SetOf<A> {
  _store: A[];

  add(a: A) {
    this._store.push(a);
  }

  transform<B>(transformer: (a: SetOf<A>) => SetOf<B>): SetOf<B> {
    return transformer(this);
  }

  forEach(fn: (a: A, index: number) => void) {
      this._store.forEach((a, i) => fn(a, i));
  }
}

function compose<A, B, C, D, E>(
  fnA: (a: SetOf<A>) => SetOf<B>, 
  fnB: (b: SetOf<B>) => SetOf<C>, 
  fnC: (c: SetOf<C>) => SetOf<D>,
  fnD: (c: SetOf<D>) => SetOf<E>,
):(x: SetOf<A>) => SetOf<E>;
/* ... etc ... */
function compose<T>(...fns: ((x: T) => T)[]): (x: T) => T {
  return (x: T) => fns.reduce((prev, fn) => fn(prev), x);
}

function map<A, B>(fn: (a: A) => B): (s: SetOf<A>) => SetOf<B> {
  return (a: SetOf<A>) => {
    const b: SetOf<B> = new SetOf();
    a.forEach(x => b.add(fn(x)));
    return b;
  }
}

function filter<A>(predicate: (a: A) => boolean): (s: SetOf<A>) => SetOf<A> {
  return (a: SetOf<A>) => {
    const result = new SetOf<A>();
    a.forEach(x => {
      if (predicate(x)) result.add(x);
    });
   return result;
  }
}

const testSet = new SetOf();
testSet.add(1);
testSet.add(2);
testSet.add(3);

// THE PROBLEM IS HERE
// all functions below are unable to infer any types.
// The user will be required to annotate the types manually at each step.
testSet.transform(
  compose(
    filter(x => x % 1 === 0),
    map(x => x + x),
    map(x => x + '!!!'),
    map(x => x.toUpperCase())
  )
)

testSet.transform(
  compose(
    filter(x => x % 1 === 0),
    map(x => x + x),
    map(x => 123),  // Whoops a bug
    map(x => x.toUpperCase()) // causes an error!
  )
)
```

### RxJS

We want to move RxJS over to using "lettable operators". That means operators that are built in a similar fashion to what you see in the example able. The current prototype augmentation is untenable for large applications, and makes tree-shaking hard if not impossible with regards to Rx.

### Ramda

A widely popular functional programming library will undoubtedly have the same issues. Especially give that (I think) the `compose` function shown above exists (probably in a better form) within Ramdba.

### Redux

`combineReducers` in redux is liable to have this same problem, at least when dealing with typed reducers. I don't think the reducer's types can be inferred inside of that call. However, I'm not sure it's a common use case for Redux to have inline reducers in a `combineReducers` call.

### Plain JavaScript

This problem would exist for any higher-order function used in this manner within JavaScript:

```ts
// given the same `compose` function from above:

const result = [1, 2, 3, 4].map(
  compose(
    x => x + x,
    x => x + '!!!',
    x => parseInt(x)
  )
);

console.log(result); // [2, 4, 6, 8]
```

cc/ @rkirov @igorminar @david-driscoll @alexeagle

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we improve type argument inference by inferring from the contextual type of a generic function call to the return type of the generic function. For example:

```ts
function emptyArray<T>(): T[] {
    return [];
}

let a1: string[] = emptyArray();  // string inferred for T
let a2: number[] = emptyArray();  // number inferred for T
```

Previously the two assignments above would error because `{}` was inferred for `T`. We now infer from the contextual type (i.e. the type of the variable to which the function result is assigned).

A more elaborate example:

```ts
function arrayFilter<T>(f: (x: T) => boolean): (a: T[]) => T[] {
    return a => a.filter(f);
}

function arrayMap<T, U>(f: (x: T) => U): (a: T[]) => U[] {
    return a => a.map(f);
}

function compose<A, B, C>(f: (x: A) => B, g: (x: B) => C): (x: A) => C {
    return x => g(f(x));
}

const positives: (a: number[]) => number[] = arrayFilter(x => x >= 0);
const lengths: (a: string[]) => number[] = arrayMap(s => s.length);
const evenLengths: (a: string[]) => number[] =
    compose(arrayMap(s => s.length), arrayFilter(x => x % 2 === 0));

let a1 = positives([1, 2, -3, 4, -5]);            // [1, 2, 4]
let a2 = lengths(['a', 'bb', 'ccc', 'dddd']);     // [1, 2, 3]
let a3 = evenLengths(['a', 'bb', 'ccc', 'dddd']); // [2, 4]
```

Inferences made from generic function return types have lower priority than inferences made from arguments to the generic function. For example, the type of `s` is inferred as `string` (not `Object`) in the following:

```ts
function filter<T>(a: T[], f: (x: T) => boolean): T[] {
    return a.filter(f);
}

let a: Object[] = filter(['a', 'bb', 'ccc', 'dddd'], s => s.length > 2);
```

This PR is a precursor for inferring higher order function types when no inferences can be made for one or more type parameters.

```ts
const wrapper = compose(x => [x], y => ({ p: y }));  // (x: {}) => { p: {}[] }
```

We currently infer `(x: {}) => { p: {}[] }`, but ideally we'd infer `<A>(x: A) => { p: A[] }`. That's next on the list.

Fixes #15680.

### CODE REVIEW

**Comment by benlesh** [comment]
I was told "if you get Anders interested in your typings issue, it'll be implemented in a few hours" 

Apparently the legend is real.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10288,27 +10293,24 @@ namespace ts {
                     // it as an inference candidate. Hopefully, a better candidate will come along that does
                     // not contain anyFunctionType when we come back to this argument for its second round
                     // of inference.
-                    if (source.flags & TypeFlags.ContainsAnyFunctionType) {
+                    if (source.flags & TypeFlags.ContainsAnyFunctionType || source === silentNeverType) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: worth adding a note on the use of silentNeverType as a marker here.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10288,27 +10293,24 @@ namespace ts {
...
-                                // T | string[], we really don't know if we should be inferring to T or not (because
-                                // the correct constituent on the target side could be string[]). Therefore, we put
-                                // such inferior inferences into a secondary bucket, and only use them if the primary
-                                // bucket is empty.
-                                const candidates = inferiority ?
-                                    inferences.secondary || (inferences.secondary = []) :
-                                    inferences.primary || (inferences.primary = []);
-                                if (!contains(candidates, source)) {
-                                    candidates.push(source);
+                    for (const inference of inferences) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: `getInferenceInfoForType` here as well?

**Comment by RyanCavanaugh** [comment]
Why doesn't this fix #11152 ?

**Comment by mhegazy** [comment]
> Why doesn't this fix #11152 ?

contextual type is not propagated/instantiated correctly yet.

**Comment by tycho01** [comment]
@ahejlsberg:
> We currently infer `(x: {}) => { p: {}[] }`, but ideally we'd infer `<A>(x: A) => { p: A[] }`. That's next on the list.

If that's still outstanding, might there be any PR/branch we could follow on that?

**Comment by johnnyreilly** [comment]
> contextual type is not propagated/instantiated correctly _**yet**_.

@mhegazy - I'm liking the use of the word "yet" above 👍 😄

**Comment by AlexGalays** [comment]
Where can we track progress against the following bug?

```ts

type Function1<T1, R> = (t1: T1) => R


function pipe<T1, R>(fn: Function1<T1, R>): Function1<T1, R> {
    return fn
}


/// test

function identity<A>(a: A): A { return a }


const pipedIdentity = pipe(identity)

// x is a {}, not a number
const x = pipedIdentity(33)
```

**Comment by gcnew** [comment]
@AlexGalays see https://github.com/Microsoft/TypeScript/issues/9366 and the linked issues.

**Pull request merged**